File: TeamCode/src/main/java/expo/util/ExpoGamepad.java
package expo.util;

import com.qualcomm.robotcore.hardware.Gamepad;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ExpoGamepad {
    private Gamepad gamepad;
    private Map<Button, ButtonToggle> toggles = new HashMap<>();
    public ExpoGamepad(Gamepad gamepad) {
        this.gamepad = gamepad;
    }

    //if someone can think of a better way to do this
    //by all means go ahead
    //i kinda dont like using hashmaps but only way i can think of
    public void update() {
        for (Map.Entry<Button, ButtonToggle> entry : toggles.entrySet()) {
            ButtonToggle toggle = entry.getValue();
            toggle.update();
        }
    }

    public void registerToggle(Button button) {
        registerToggle(button, false);
    }
    public void registerToggle(Button button, boolean state) {
        toggles.put(button, new ButtonToggle(this,button, state));
    }

    public boolean getToggle(Button button) {
        ButtonToggle toggle = toggles.get(button);
        if(toggle == null)
            return false;
        return toggle.getValue();
    }


    public boolean getButton(Button button) {
        switch (button) {
            case A:
                return gamepad.a;
            case B:
                return gamepad.b;
            case X:
                return gamepad.x;
            case Y:
                return gamepad.y;
            case START:
                return gamepad.start;
            case BACK:
                return gamepad.back;
            case DOWN:
                return gamepad.dpad_down;
            case LEFT:
                return gamepad.dpad_left;
            case RIGHT:
                return gamepad.dpad_right;
            case UP:
                return gamepad.dpad_up;
            case LEFT_BUMPER:
                return gamepad.left_bumper;
            case RIGHT_BUMPER:
                return gamepad.right_bumper;
            case LEFT_STICK:
                return gamepad.left_stick_button;
            case RIGHT_STICK:
                return gamepad.right_stick_button;
        }
        //this will never run and if it does u have bigger problems
        return false;
    }

    public double getLeftX() {
        return gamepad.left_stick_x;
    }

    public double getLeftY() {
        return gamepad.left_stick_y;
    }

    public double getRightX() {
        return gamepad.right_stick_x;
    }

    public double getRightY() {
        return gamepad.right_stick_y;
    }
}

File: TeamCode/src/main/java/expo/util/ExpoGamepad.kt
package expo.util

import com.qualcomm.robotcore.hardware.Gamepad
import java.util.*

class ExpoGamepad(private val gamepad: Gamepad) {
    var toggles = HashMap<Button, ButtonToggle>()

    fun update() {
        for (entry: Map.Entry<Button, ButtonToggle> in toggles.entries) {
            val toggle: ButtonToggle = entry.value
            toggle.update()
        }
    }

    fun registerToggle(button: Button) {
        registerToggle(button, false)
    }

    fun registerToggle(button: Button, state: Boolean) {
        toggles[button] = ButtonToggle(this, button, state)
    }

    fun getToggle(button: Button): Boolean {
        val toggle: ButtonToggle = toggles[button] ?: return false
        return toggle.getValue()
    }

    fun getButton(button: Button): Boolean {
        return when (button) {
            Button.A -> gamepad.a && !gamepad.start
            Button.B -> gamepad.b && !gamepad.start
            Button.X -> gamepad.x
            Button.Y -> gamepad.y
            Button.START -> gamepad.start
            Button.BACK -> gamepad.back
            Button.DOWN -> gamepad.dpad_down
            Button.LEFT -> gamepad.dpad_left
            Button.RIGHT -> gamepad.dpad_right
            Button.UP -> gamepad.dpad_up
            Button.LEFT_BUMPER -> gamepad.left_bumper
            Button.RIGHT_BUMPER -> gamepad.right_bumper
            Button.LEFT_STICK -> gamepad.left_stick_button
            Button.RIGHT_STICK -> gamepad.right_stick_button
        }
    }

    fun getLeftX(): Double {
        return gamepad.left_stick_x.toDouble()
    }

    fun getLeftY(): Double {
        return gamepad.left_stick_y.toDouble()
    }

    fun getRightX(): Double {
        return gamepad.right_stick_x.toDouble()
    }

    fun getRightY(): Double {
        return gamepad.right_stick_y.toDouble()
    }

    fun getControllerVector(): Vector = Vector(getLeftX(), getLeftY())

    fun getControllerVector(angle: Double): Vector {
        val controlVector = Vector(getLeftX(), getLeftY())
        controlVector.rotate(angle)
        return controlVector
    }
}